File: apps/web/lib/api/installments.ts
import 'server-only';

import { authenticatedFetch } from '@/lib/authenticated-fetch';

// --- Raw types (API JSON shape, snake_case) ---

interface InstallmentRaw {
  id: number;
  name: string;
  total_amount: string;
  installment_amount: string;
  currency: string;
  converted_total_amount: string | null;
  converted_installment_amount: string | null;
  installments_count: number;
  current_installment: number;
  start_date: string;
  next_cuota_date: string | null;
  payment_method: string | null;
  credit_card_id: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// --- Frontend types (camelCase) ---

export interface Installment {
  id: number;
  name: string;
  totalAmount: string;
  installmentAmount: string;
  currency: string;
  convertedTotalAmount: string | null;
  convertedInstallmentAmount: string | null;
  installmentsCount: number;
  currentInstallment: number;
  startDate: string;
  // Derived from `start_date + (current_installment - 1) months` server-side so the
  // installments table can render the next-due-date in the column where subscriptions
  // show `next_billing_date` and obligations show `next_due_date`. Null when the plan
  // is fully paid (`current_installment > installments_count`).
  nextCuotaDate: string | null;
  paymentMethod: string | null;
  creditCardId: number | null;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

// --- Mappers ---

function mapInstallment(raw: InstallmentRaw): Installment {
  return {
    id: raw.id,
    name: raw.name,
    totalAmount: raw.total_amount,
    installmentAmount: raw.installment_amount,
    currency: raw.currency,
    convertedTotalAmount: raw.converted_total_amount,
    convertedInstallmentAmount: raw.converted_installment_amount,
    installmentsCount: raw.installments_count,
    currentInstallment: raw.current_installment,
    startDate: raw.start_date,
    nextCuotaDate: raw.next_cuota_date,
    paymentMethod: raw.payment_method,
    creditCardId: raw.credit_card_id,
    isActive: raw.is_active,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

// --- API functions ---

export type InstallmentSortField =
  | 'name'
  | 'total_amount'
  | 'installment_amount'
  | 'currency'
  | 'installments_count'
  | 'current_installment'
  | 'start_date'
  | 'next_cuota_date';
export type SortOrder = 'asc' | 'desc';

export interface GetInstallmentsParams {
  search?: string;
  sortBy?: InstallmentSortField;
  sortOrder?: SortOrder;
  showArchived?: boolean;
  // Archived plan ids to include alongside the active set — used by the expense edit
  // dialog so a row linked to a since-archived plan can still surface in the linked-FK
  // dropdown. Server ignores when showArchived=true.
  includeIds?: number[];
  currency?: string;
}

export async function getInstallments(params: GetInstallmentsParams = {}): Promise<Installment[]> {
  const qs = new URLSearchParams();
  if (params.search) qs.set('search', params.search);
  if (params.sortBy) qs.set('sort_by', params.sortBy);
  if (params.sortOrder) qs.set('sort_order', params.sortOrder);
  if (params.showArchived) qs.set('show_archived', 'true');
  if (params.includeIds && params.includeIds.length > 0) {
    for (const id of params.includeIds) qs.append('include_ids', String(id));
  }
  if (params.currency) qs.set('currency', params.currency);

  const endpoint = qs.toString() ? `/installments?${qs.toString()}` : '/installments';
  const res = await authenticatedFetch(endpoint, { method: 'GET' });
  if (!res.ok) throw new Error('Failed to fetch installments');
  const raw: InstallmentRaw[] = await res.json();
  return raw.map(mapInstallment);
}

File: apps/api/app/routers/installments.py
from fastapi import APIRouter, Query, status

from app.deps.api_key_auth import JwtOrApiKeyUser
from app.deps.auth import CurrentUser
from app.deps.db import SessionDep
from app.schemas.installment import InstallmentCreate, InstallmentResponse, InstallmentUpdate
from app.services import installment_service

router = APIRouter(prefix="/installments", tags=["installments"])


# List installments for the current user with optional search, sorting, and currency conversion.
@router.get("", response_model=list[InstallmentResponse])
async def list_installments(
    current_user: CurrentUser,
    session: SessionDep,
    search: str | None = Query(default=None, description="Filter installments by name (case-insensitive)."),
    sort_by: str | None = Query(
        default=None,
        description=(
            "Column to sort by (name, total_amount, installment_amount, currency, "
            "installments_count, current_installment, start_date, next_cuota_date)."
        ),
    ),
    sort_order: str = Query(default="asc", description="Sort direction (asc or desc)."),
    show_archived: bool = Query(default=False, description="Include archived (completed) installments."),
    include_ids: list[int] | None = Query(
        default=None,
        description=(
            "Archived plan ids to include alongside the active set. Used by the expense edit dialog "
            "so a row linked to a since-archived plan can still render the plan name in the dropdown. "
            "Ignored when show_archived=true (everything is already included)."
        ),
    ),
    currency: str | None = Query(default=None, description="Display currency (e.g. USD, ARS). Omit for original."),
) -> list[InstallmentResponse]:
    return await installment_service.list_installments(
        session,
        current_user,
        search=search,
        sort_by=sort_by,
        sort_order=sort_order,
        active_only=not show_archived,
        include_ids=include_ids,
        currency=currency,
    )


# Get a single installment by id (with optional currency conversion).
@router.get("/{installment_id}", response_model=InstallmentResponse)
async def get_installment(
    installment_id: int,
    current_user: CurrentUser,
    session: SessionDep,
    currency: str | None = Query(default=None, description="Display currency (e.g. USD, ARS). Omit for original."),
) -> InstallmentResponse:
    return await installment_service.get_installment_response(session, installment_id, current_user, currency=currency)


# Create a new installment plan. Supports both JWT (web) and API key (iOS Shortcut) auth.
@router.post("", response_model=InstallmentResponse, status_code=status.HTTP_201_CREATED)
async def create_installment(
    body: InstallmentCreate,
    current_user: JwtOrApiKeyUser,
    session: SessionDep,
) -> InstallmentResponse:
    installment = await installment_service.create_installment(
        session,
        current_user,
        name=body.name,
        total_amount=body.total_amount,
        installment_amount=body.installment_amount,
        currency=body.currency,
        installments_count=body.installments_count,
        current_installment=body.current_installment,
        start_date=body.start_date,
        payment_method=body.payment_method,
        credit_card_id=body.credit_card_id,
    )
    return InstallmentResponse.model_validate(installment)


# Update an existing installment plan.
@router.put("/{installment_id}", response_model=InstallmentResponse)
async def update_installment(
    installment_id: int,
    body: InstallmentUpdate,
    current_user: CurrentUser,
    session: SessionDep,
) -> InstallmentResponse:
    payload = body.model_dump(exclude_unset=True)
    installment = await installment_service.update_installment(session, installment_id, current_user, **payload)
    return InstallmentResponse.model_validate(installment)


# Delete an installment plan. Returns 204.
@router.delete("/{installment_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_installment(
    installment_id: int,
    current_user: CurrentUser,
    session: SessionDep,
) -> None:
    await installment_service.delete_installment(session, installment_id, current_user)
